applications/sdp/mspi/src/hrt/hrt.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include "hrt.h"
+#include <hal/nrf_vpr_csr_vio.h>
+#include <hal/nrf_vpr_csr_vtim.h>
+
+#define TOP 4
+
+void write_single_by_word(volatile uint32_t *data, uint8_t data_len, uint32_t counter_top,
+			  uint8_t word_size, bool ce_enable_state, bool hold_ce)
+{
+	NRFX_ASSERT(word_size > MAX_WORD_SIZE);
+
+	/* Configuration step */
+	uint16_t dir = nrf_vpr_csr_vio_dir_get();
+
+	nrf_vpr_csr_vio_dir_set(dir | PIN_DIR_OUT_MASK(D0_PIN) | PIN_DIR_OUT_MASK(CS_PIN) |
+				PIN_DIR_OUT_MASK(SCLK_PIN));
+
+	uint16_t out = nrf_vpr_csr_vio_out_get();
+
+	nrf_vpr_csr_vio_out_set(out | PIN_OUT_LOW_MASK(D0_PIN) | PIN_OUT_HIGH_MASK(CS_PIN) |
+				PIN_OUT_LOW_MASK(SCLK_PIN));
+
+	nrf_vpr_csr_vio_mode_out_t out_mode = {
+		.mode = NRF_VPR_CSR_VIO_SHIFT_OUTB_TOGGLE,
+		.frame_width = 1,
+	};
+
+	nrf_vpr_csr_vio_mode_out_set(&out_mode);
+	nrf_vpr_csr_vio_mode_in_buffered_set(NRF_VPR_CSR_VIO_MODE_IN_CONTINUOUS);
+
+	nrf_vpr_csr_vio_config_t config;
+
+	nrf_vpr_csr_vio_config_get(&config);
+	config.input_sel = false;
+	nrf_vpr_csr_vio_config_set(&config);
+
+	/* Fix position of data if word size < MAX_WORD_SIZE,
+	 * so that leading zeros would not be printed instead of data bits.
+	 */
+	if (word_size < MAX_WORD_SIZE) {
+		for (uint8_t i = 0; i < data_len; i++) {
+			data[i] = data[i] << (MAX_WORD_SIZE - word_size);
+		}
+	}
+
+	/* Counter settings */
+	nrf_vpr_csr_vtim_count_mode_set(0, NRF_VPR_CSR_VTIM_COUNT_RELOAD);
+	nrf_vpr_csr_vtim_simple_counter_top_set(0, counter_top);
+
+	/* Set number of shifts before OUTB needs to be updated.
+	 * First shift needs to be increased by 1.
+	 */
+	nrf_vpr_csr_vio_shift_cnt_out_set(word_size);
+	nrf_vpr_csr_vio_shift_cnt_out_buffered_set(word_size - 1);
+
+	/* Enable CS */
+	out = nrf_vpr_csr_vio_out_get();
+	out &= ~PIN_OUT_HIGH_MASK(CS_PIN);
+	out |= ce_enable_state ? PIN_OUT_HIGH_MASK(CS_PIN) : PIN_OUT_LOW_MASK(CS_PIN);
+	nrf_vpr_csr_vio_out_set(out);
+
+	/* Start counter */
+	nrf_vpr_csr_vtim_simple_counter_set(0, 3 * counter_top);
```

Comment:
3 should be a define

Reply:
3 is a temporary value that will be fixed when baudrate configuration will be implemented

Reply:
I would still move it to a define, since its value will most likely be different for other platforms (for example, H20)

Reply:
ok, fixed